populator-machinery/controller_test.go

```diff
@@ -0,0 +1,538 @@
+	kubeClient.CoreV1().Pods(testVpWorkingNamespace).Delete(context.TODO(), testPodName, metav1.DeleteOptions{})
+	kubeClient.CoreV1().PersistentVolumes().Delete(context.TODO(), testPvName, metav1.DeleteOptions{})
+}
+
+func compareResult(want error, got error) bool {
+	if want == nil {
+		return got == nil
+	}
+	if got == nil {
+		return want == nil
+	}
+	return want.Error() == got.Error()
+}
+
+func compareNotifyMap(want []string, got map[string]*stringSet) error {
+	if len(want) != len(got) {
+		return fmt.Errorf("The number of keys expected is different from actual. Expect %v, got %v", len(want), len(got))
+	}
+	for _, key := range want {
+		if got[key] == nil {
+			return fmt.Errorf("Expected key %s not found in the notifyMap", key)
+		}
+	}
+	return nil
+}
+
+func runSyncPvcTests(tests []testCase, t *testing.T) {
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			for _, obj := range test.initialObjects {
+				switch obj.(type) {
+				case *v1.PersistentVolumeClaim:
+					pvc := obj.(*v1.PersistentVolumeClaim)
+					_, err := kubeClient.CoreV1().PersistentVolumeClaims(pvc.ObjectMeta.Namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+					if err != nil {
+						t.Fatalf("Create pvc failed: %s", err.Error())
+					}
+					pvcInformer.Informer().GetStore().Add(obj)
+				case *unstructured.Unstructured:
+					unstInformer.GetStore().Add(obj)
+				case *storagev1.StorageClass:
+					scInformer.Informer().GetStore().Add(obj)
+				case *v1.Pod:
+					pod := obj.(*v1.Pod)
+					_, err := kubeClient.CoreV1().Pods(pod.ObjectMeta.Namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
+					if err != nil {
+						t.Fatalf("Create pod failed: %s", err.Error())
+					}
+					podInformer.Informer().GetStore().Add(obj)
+				case *v1.PersistentVolume:
+					pv := obj.(*v1.PersistentVolume)
+					_, err := kubeClient.CoreV1().PersistentVolumes().Create(context.TODO(), pv, metav1.CreateOptions{})
+					if err != nil {
+						t.Fatalf("Create pv failed: %s", err.Error())
+					}
+					pvInformer.Informer().GetStore().Add(obj)
+				default:
+					t.Fatalf("Unknown initalObject type: %+v", obj)
+				}
+			}
+
+			c := initController()
+			result := c.syncPvc(context.TODO(), test.key, test.pvcNamespace, test.pvcName)
+			if !compareResult(test.expectedResult, result) {
+				t.Errorf("Error: expected result %t, got %t", test.expectedResult, result)
+			}
+			err := compareNotifyMap(test.expectedKeys, c.notifyMap)
+			if err != nil {
+				t.Errorf(err.Error())
+			}
+			cleanup()
+		})
+	}
+}
+
+func TestSyncPvc(t *testing.T) {
+	dataSourceKey := "unstructured/" + testPvcNamespace + "/" + testDataSourceName
+	storageClassKey := "sc/" + testStorageClassName
+	podKey := "pod/" + testVpWorkingNamespace + "/" + testPodName
+	pvcPrimeKey := "pvc/" + testVpWorkingNamespace + "/" + testPopulatorPvcName
+	pvKey := "pv/" + testPvName
+
+	tests := []testCase{
+		{
+			name:         "Ignore PVCs in working namespace",
```

Comment:
nit: Ignore PVCs in controller's working namespace

Reply:
Done